src/controllers/auth.controller.ts

```diff
@@ -0,0 +1,158 @@
+import admin from 'firebase-admin';
+import { Request, Response, NextFunction } from 'express';
+
+import * as ERR_MSG from '../constants/ErrResponses';
+
+
+/*
+  Create a new auth object for the provided email and password.
+  Also links the auth account with existing user document and
+  updates document with provided profile information.
+  Prereqs:
+    1.  Request body contains the following key value pairs:
+        a. email
+        b. password
+        c. first_name
+        d. last_name
+        e. major
+        f. grad_year
+  Return: On successful account creation, user document is updated with incoming profile
+          information and user auth account object is sent back mapped to newUser.
+  Errors:`Can send back the following error responses: 400, 403, 500.
+*/
+export const signup = async (
+  req: Request,
+  res: Response,
+  next: NextFunction
+): Promise<void> => {
+
+  // Grab form information from request body or if body is undefined,
+  // parse from {}. (Helps avoid TypeError)
+  const { email,
+    password,
+    first_name,
+    last_name,
+    major,
+    graduation_year } = req.body || {};
+
+  if (!validInput(email, password, first_name, last_name, major, graduation_year)) {
+    // Invalid user information
+    res.status(400).json({
+      msg: ERR_MSG.GENERIC_400_MSG,
+    });
+    return next();
+  }
+
+  const userDocID = await getUserDocFromEmail(email);
+
+  if (!userDocID) {
+    // email is not whitelisted.
+    res.status(403).json({
+      msg: ERR_MSG.USER_NOT_WHITELISTED,
+    });
+    return next();
+  }
+
+  const newUser: admin.auth.UserRecord = await admin.auth().createUser({
+    uid: userDocID,
+    email: email,
+    password: password
+  })
+    .then(userRec => {
+      return userRec;
+    })
+    .catch(err => {
+      if (typeof (err.code.includes) === 'function' && err.code.includes('exists')) {
```

Comment:
not a great way of validating error codes since includes might cause problems down the line; if firestore auth has error codes as types I'd rather we explicitly list which codes we're catching